### ISSUE TITLE
ArrayIndexOutOfBoundsException

### ISSUE THREAD

**Issue by diamondq** [wildfly/jandex#80]
Ran into a case where scanning some libraries found an ArrayIndexOutOfBoundsException. This was based on 2.1.3.Final.

```
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 1
	at org.jboss.jandex.Indexer.updateTypeTarget(Indexer.java:841)
	at org.jboss.jandex.Indexer.updateTypeTargets(Indexer.java:624)
	at org.jboss.jandex.Indexer.index(Indexer.java:1603)
	at test.Main.main(Main.java:14)
```
This occurred using a simple test program:
```
package test;

import java.io.IOException;
import java.io.InputStream;

import org.jboss.jandex.Index;
import org.jboss.jandex.Indexer;

public class Main {

  public static void main(String[] args) throws IOException {
    Indexer indexer = new Indexer();
    InputStream stream = Main.class.getClassLoader().getResourceAsStream("test/Broken.class");
    indexer.index(stream);
    Index index = indexer.complete();
  }
}
```
and failed on this class:
```
package test;

import java.util.List;

import org.checkerframework.checker.nullness.qual.NonNull;

public class Broken {

  public static <U extends @NonNull List<String>> U listOf() {
    return null;
  }

}
```
This was, obviously, from a much more complicated example, but I tried to reduce it down to the simplest case that still generated the exception.

As far as I could tell, the problem lies here:
```
            case TYPE_PARAMETER_BOUND: {
                TypeParameterBoundTypeTarget boundTarget = target.asTypeParameterBound();
                type = getTypeParameters(enclosingTarget)[boundTarget.position()]
                           .asTypeVariable().boundArray()[boundTarget.boundPosition()];
                break;
            }
```
where the boundTarget.position is 1, and boundPosition is 1, but while the getTypeParameters returns a 2-element array, the second array (for boundPosition) is only a 1-element array, and thus, fails.
Unfortunately, I couldn't follow the code logic enough to track the underlying problem, so just reporting with all the data I did get.

Cheers.

**Comment by vlsi** [comment]
>Unfortunately, I couldn't follow the code logic enough to track the underlying problem, so just reporting with all the data I did get.

I explained the issue in https://github.com/wildfly/jandex/issues/99
The key point is that `U extends @NonNull List<String>` compiles to `U::List...` signature, so the bound indexing starts with 1, and the position of 0 is omitted.

That is why `[boundTarget.boundPosition()]` is not valid.

**Comment by n1hility** [comment]
This should be fixed now in 2.2.3
